spread-shellcheck

```diff
@@ -161,20 +162,21 @@ def checkfile(path):
 
     if path.endswith('spread.yaml') and 'suites' in data:
         # check suites
-        with ThreadPoolExecutor() as executor:
-            futures = []
-            for suite in data['suites'].keys():
-                for section in SECTIONS:
-                    if section not in data['suites'][suite]:
-                        continue
-                        logging.debug("%s (suite %s): checking section %s", path, suite, section)
-                        futures.append(executor.submit(checksection, data['suites'][suite][section], env))
-            for future in futures:
-                try:
-                    future.result()
-                except ShellcheckRunError as serr:
-                    errors.addfailure('suites/' + suite + '/' + section,
-                                    serr.stderr.decode('utf-8'))
+        suites_sections_and_futures = []
+        for suite in data['suites'].keys():
+            for section in SECTIONS:
+                if section not in data['suites'][suite]:
+                    continue
+                logging.debug("%s (suite %s): checking section %s", path, suite, section)
+                future = executor.submit(checksection, data['suites'][suite][section], env)
+                suites_sections_and_futures.append((suite, section, future))
+        for item in suites_sections_and_futures:
```

Comment:
maybe we could use this so that we are collecting results as they come rather than being blocked on the slow(er) ones:
```

for suite in ..
    for section in SECTIONS:
        ...
        future = future.submit(...)
        futures_to_items[future] = (suite, section)

for ft in futures.as_completed(all_futures):
    suite, section = futures_to_items[ft]
    try:
       ft.result()
    except ShellcheckRunnError as serr:
        ...
```

Reply:
Yes but I'd like to do this in a follow up to let a correct version land once

Reply:
Waiting on other futures from within a job also introduces the possibility of deadlocks, although it is mostly theoretical for how the code is being called.

Imagine that we filled up the threadpool with jobs with `spread.yaml` style jobs, and then each of those ends up waiting for other futures to complete.  One alternative would be to not run `checkfile` in the executor, but instead have it return a list of (section, future) pairs.

Then it's just a matter of collecting the futures from all the `checkfile` invocations and waiting on them.

Reply:
Yeah, that's true. We could change the code to yield all the TODO items and send it off to a farm of processes or threads to execute. Having said that, I think what we have is fine for now :)